core/Tracker/Failures.php

```diff
@@ -0,0 +1,178 @@
+use Piwik\Site;
+use Piwik\Db as PiwikDb;
+
+class Failures
+{
+    const CLEANUP_OLD_FAILURES_DAYS = 2;
+    const FAILURE_ID_INVALID_SITE = 1;
+    const FAILURE_ID_NOT_AUTHENTICATED = 2;
+
+    private $table = 'tracking_failure';
+    private $tablePrefixed;
+    private $now;
+
+    public function __construct()
+    {
+        $this->tablePrefixed = Common::prefixTable($this->table);
+    }
+
+    public function setNow(Date $now)
+    {
+        $this->now = $now;
+    }
+
+    private function getNow()
+    {
+        if (isset($this->now)) {
+            return $this->now;
+        }
+        return Date::now();
+    }
+
+    public function logFailure($idFailure, Request $request)
+    {
+        $visitExcluded = new VisitExcluded($request);
+        if ($visitExcluded->isExcluded()) {
+            return;
+        }
+
+        $idSite = (int) $request->getIdSite();
+        $idFailure = (int) $idFailure;
+
+        if ($idSite > 9999999 || $idSite < 0 || $this->hasLoggedFailure($idSite, $idFailure)) {
+            return; // we prevent creating huge amount of entries in the cache
+        }
+
+        $params = $this->getParamsWithTokenAnonymized($request);
+        $sql = sprintf('INSERT INTO %s (`idsite`, `idfailure`, `date_first_occurred`, `request_url`) VALUES(?,?,?,?) ON DUPLICATE KEY UPDATE idsite=idsite;', $this->tablePrefixed);
+
+        PiwikDb::get()->query($sql, array($idSite, $idFailure, $this->getNow()->getDatetime(), http_build_query($params)));
+    }
+
+    private function hasLoggedFailure($idSite, $idFailure)
+    {
+        $sql = sprintf('SELECT idsite FROM %s WHERE idsite = ? and idfailure = ?', $this->tablePrefixed);
+        $row = PiwikDb::fetchRow($sql, array($idSite, $idFailure));
+
+        return !empty($row);
+    }
+
+    private function getParamsWithTokenAnonymized(Request $request)
+    {
+        // eg if there is a typo in the token auth we want to replace it as well to not accidentally leak a token
+        // eg imagine a super user tries to issue an API request for a site and sending the wrong parameter for a token...
+        // an admin may have view access for this and can see the super users token
+        $token = $request->getTokenAuth();
+        $params = $request->getRawParams();
+        foreach (array('token_auth', 'token', 'tokenauth', 'token__auth') as $key) {
+            if (isset($params[$key])) {
+                $params[$key] = '__ANONYMIZED__';
+            }
+        }
+        foreach ($params as $key => $value) {
+            if (!empty($token) && $value === $token) {
+                $params[$key] = '__ANONYMIZED__'; // user accidentally posted the token in a wrong field
+            } elseif (!empty($value) && is_string($value) && Common::mb_strlen($value) >= 29 && Common::mb_strlen($value) <= 36 && ctype_xdigit($value)) {
+                $params[$key] = '__ANONYMIZED__'; // user maybe posted a token in a different field... it looks like it might be a token
+            }
```

Comment:
Would it be simpler to do `str_replace($token, '__ANONYMIZED__', $value)` on each param?

Also I think __ANONYMIZED__ might be confusing to some people, since requests are also anonymized by changing IP addresses. Maybe `__TOKEN_AUTH__` might be clearer?

Reply:
Will change it to `__TOKEN_AUTH__ `. didn't use only `str_replace($token, '__ANONYMIZED__', $value)` in case there were any typos or just one character missing or one too much.

Reply:
I see, 👍